rbac/api/cross_access/relation_api_dual_write_cross_access_handler.py

```diff
@@ -0,0 +1,175 @@
+        try:
+            self._replicator.replicate(
+                ReplicationEvent(
+                    event_type=self.event_type,
+                    info={"user_uuid": str(self.cross_account_principal.user_id)},
+                    partition_key=PartitionKey.byEnvironment(),
+                    remove=self.relations_to_remove,
+                    add=self.relations_to_add,
+                ),
+            )
+        except Exception as e:
+            raise DualWriteException(e)
+
+    def _create_default_mapping_for_system_role(self, system_role: Role):
+        """Create default mapping."""
+        assert system_role.system is True, "Expected system role. Mappings for custom roles must already be created."
+        binding = V2rolebinding(
+            str(uuid4()),
+            # Assumes same role UUID for V2 system role equivalent.
+            V2role.for_system_role(str(system_role.uuid)),
+            V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
+            users=frozenset([str(self.cross_account_principal.user_id)]),
+        )
+        mapping = BindingMapping.for_role_binding(binding, system_role)
+        self.relations_to_add.extend(mapping.as_tuples())
+        return mapping
+
+    def generate_relations_to_add_roles(self, roles: Iterable[Role]):
+        """Generate relations to add roles."""
+        if not self.replication_enabled():
+            return
+
+        def add_principal_to_binding(mapping: BindingMapping):
+            self.relations_to_add.append(mapping.add_user_to_bindings(str(self.cross_account_principal.user_id)))
+
+        for role in roles:
+            self._update_mapping_for_system_role(
+                role,
+                update_mapping=add_principal_to_binding,
+                create_default_mapping_for_system_role=lambda: self._create_default_mapping_for_system_role(role),
+            )
+
+    def replicate(self):
+        """Replicate added role."""
+        if not self.replication_enabled():
+            return
+
+        self._replicate()
+
+    def generate_relations_to_remove_roles(self, roles: Iterable[Role]):
+        """Replicate removed role."""
+        if not self.replication_enabled():
+            return
+
+        def remove_principal_from_binding(mapping: BindingMapping):
+            self.relations_to_remove.append(
+                mapping.remove_user_from_bindings(str(self.cross_account_principal.user_id))
+            )
+
+        for role in roles:
+            self._update_mapping_for_system_role(
+                role, update_mapping=remove_principal_from_binding, create_default_mapping_for_system_role=lambda: None
+            )
+
+    def _update_mapping_for_system_role(
+        self,
+        role: Role,
+        update_mapping: Callable[[BindingMapping], None],
+        create_default_mapping_for_system_role: Callable[[], Optional[BindingMapping]],
+    ):
+        assert role.system is True, "Expected system role."
```

Comment:
It's probably worth this being an if check and raising an exception in this case just because it would be really bad if it somehow wasn't a system role (e.g. through some path that is missed in testing)